homeassistant/components/sensor/integration.py

```diff
@@ -0,0 +1,180 @@
+UNIT_PREFIXES = {None: 1,
+                 "k": 10**3,
+                 "G": 10**6,
+                 "T": 10**9}
+
+# SI Time prefixes
+UNIT_TIME = {'s': 1,
+             'min': 60,
+             'h': 60*60,
+             'd': 24*60*60}
+
+ICON = 'mdi:char-histogram'
+
+DEFAULT_ROUND = 3
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_SOURCE_SENSOR): cv.entity_id,
+    vol.Optional(CONF_ROUND_DIGITS, default=DEFAULT_ROUND): vol.Coerce(int),
+    vol.Optional(CONF_UNIT_PREFIX, default=None): vol.In(UNIT_PREFIXES),
+    vol.Optional(CONF_UNIT_TIME, default='h'): vol.In(UNIT_TIME),
+    vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the integration sensor."""
+    integral = IntegrationSensor(config[CONF_SOURCE_SENSOR],
+                                 config.get(CONF_NAME),
+                                 config[CONF_ROUND_DIGITS],
+                                 config[CONF_UNIT_PREFIX],
+                                 config[CONF_UNIT_TIME],
+                                 config.get(CONF_UNIT_OF_MEASUREMENT))
+
+    async_add_entities([integral])
+
+
+class IntegrationSensor(RestoreEntity):
+    """Representation of an integration sensor."""
+
+    def __init__(self, source_entity, name, round_digits, unit_prefix,
+                 unit_time, unit_of_measurement):
+        """Initialize the integration sensor."""
+        self._sensor_source_id = source_entity
+        self._round_digits = round_digits
+        self._state = 0
+
+        self._name = name if name is not None\
+            else '{} integral'.format(source_entity)
+
+        if unit_of_measurement is None:
+            self._unit_of_measurement = "" if unit_prefix is None\
+                                        else unit_prefix
+            self._unit_of_measurement += "{}" + unit_time
+        else:
+            self._unit_of_measurement = unit_of_measurement
+
+        self._unit_prefix = UNIT_PREFIXES[unit_prefix]
+        self._unit_time = UNIT_TIME[unit_time]
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            try:
+                self._state = Decimal(state.state)
+            except ValueError as err:
+                _LOGGER.warning("Could not restore last state: %s", err)
+
+        @callback
+        def calc_integration(entity, old_state, new_state):
+            """Handle the sensor state changes."""
+            if old_state is None or\
+                    old_state.state in [STATE_UNKNOWN, STATE_UNAVAILABLE] or\
+                    new_state.state in [STATE_UNKNOWN, STATE_UNAVAILABLE]:
+                return
+
+            if "{}" in self._unit_of_measurement:
```

Comment:
I think it's more clear if we save a default unit template as another attribute, and check for `None` here instead.

Reply:
good point! fixed